lib/cisco_node_utils/itd_device_group.rb

```diff
@@ -0,0 +1,210 @@
+      @get_args = @set_args
+    end
+
+    def probe_params
+      params = config_get('itd_device_group', 'probe', @get_args)
+      params
+    end
+
+    # proble configuration is all done in a single line (like below)
+    # probe tcp port 32 frequency 10 timeout 5 retry-down-count 3 ...
+    # probe udp port 23 frequency 10 timeout 5 retry-down-count 3 ...
+    # probe icmp frequency 10 timeout 5 retry-down-count 3 retry-up-count 3
+    # probe dns host 8.8.8.8 frequency 10 timeout 5 retry-down-count 3 ...
+    # also the 'control enable' can be set if the type is tcp or udp only
+    # probe udp port 23 control enable frequency 10 timeout 5 ...
+    def probe
+      params = probe_params
+      hash = {}
+      hash[:probe_control] = default_probe_control
+      if params.nil?
+        hash[:probe_frequency] = default_probe_frequency
+        hash[:probe_timeout] = default_probe_timeout
+        hash[:probe_retry_down] = default_probe_retry_down
+        hash[:probe_retry_up] = default_probe_retry_up
+        hash[:probe_type] = default_probe_type
+        return hash
+      end
+      hash[:probe_frequency] = params[1].to_i
+      hash[:probe_timeout] = params[2].to_i
+      hash[:probe_retry_down] = params[3].to_i
+      hash[:probe_retry_up] = params[4].to_i
+
+      lparams = params[0].split
+      hash[:probe_type] = lparams[0]
+      case hash[:probe_type].to_sym
+      when :dns
+        hash[:probe_dns_host] = lparams[2]
+      when :tcp, :udp
+        hash[:probe_port] = lparams[2].to_i
+        hash[:probe_control] = true unless lparams[3].nil?
+      end
+      hash
+    end
+
+    def probe_control
+      hash = probe
+      hash[:probe_control]
+    end
+
+    def default_probe_control
+      config_get_default('itd_device_group', 'probe_control')
+    end
+
+    def probe_dns_host
+      hash = probe
+      hash[:probe_dns_host]
+    end
+
+    def probe_frequency
+      hash = probe
+      hash[:probe_frequency]
+    end
+
+    def default_probe_frequency
+      config_get_default('itd_device_group', 'probe_frequency')
+    end
+
+    def probe_port
+      hash = probe
+      hash[:probe_port]
+    end
+
+    def probe_retry_down
+      hash = probe
+      hash[:probe_retry_down]
+    end
+
+    def default_probe_retry_down
+      config_get_default('itd_device_group', 'probe_retry_down')
+    end
+
+    def probe_retry_up
+      hash = probe
+      hash[:probe_retry_up]
+    end
+
+    def default_probe_retry_up
+      config_get_default('itd_device_group', 'probe_retry_up')
+    end
+
+    def probe_timeout
+      hash = probe
+      hash[:probe_timeout]
+    end
+
+    def default_probe_timeout
+      config_get_default('itd_device_group', 'probe_timeout')
+    end
+
+    def probe_type
+      hash = probe
+      hash[:probe_type]
+    end
+
+    def default_probe_type
+      config_get_default('itd_device_group', 'probe_type')
+    end
+
+    def probe=(type, host, control, freq, ret_up, ret_down, port, timeout)
```

Comment:
I would strongly recommend you follow the `ace` design pattern for the setter.  We already have a working design pattern that uses the `ace_set` method in puppet.  This allows you to simply pass all of the arguments to the setter as a hash.

- NodeUtils `ace_set(attrs)` method: https://github.com/cisco/cisco-network-node-utils/blob/develop/lib/cisco_node_utils/ace.rb#L100
- Then in puppet you can do this: https://github.com/cisco/cisco-network-puppet-module/blob/develop/lib/puppet/provider/cisco_ace/cisco.rb#L157